spi-proposal.md

```diff
@@ -0,0 +1,249 @@
+either **Read/write**` mode or **Read-only** mode is used for the `SPI_TRANSFER`.
+
+A `SPI_END` command disables the SPI bus for the specified channel.
+
+
+### SPI_CONFIG
+
+Use `SPI_CONFIG` to initialize the SPI bus. Up to 4 SPI ports are supported,
+where each port is identified by a `channel` number (0-3).
+
+`SPI_CONFIG` must be called at least once before sending any of the other
+commands.
+
+`channel` is used to identify which SPI bus is used in the case that a board
+has multiple ports (SPI, SPI1, SPI2, etc). Many only have one port so the
+`channel` value will most often be `0`.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_CONFIG           (0x00)
+3:  channel              (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### SPI_BEGIN_TRANSACTION
+
+Initialize a SPI device. Call once before reading from or writing to a separate
+SPI device.
+
+The `deviceId` is any value between 0 and 31 and is chosen by the user. The
+value should be unique for each device used in an application. The `deviceId`
+will be returned with each `SPI_REPLY` message so it is clear which device the
+data corresponds to.
+
+A chip select pin (`csPin`) can optionally be specified. This pin will be
+controlled per the rules specified in `pinOptions` byte of the `SPI_TRANSACTION`
+command. For uncommon use cases of CS or other required HW pins per certain SPI
+devices, it is better to control them separately using Firmata `DIGITAL_MESSAGE`.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_BEGIN_TRANSACTION (0x01)
+3:  deviceId | channel    (bits 2-6: deviceId, bits 0-1: channel)
+4:  dataMode | bitOrder   (bits 1-2: dataMode (0-3), bit 0: bitOrder)
+5:  maxSpeed              (bits 0 - 6)
+6:  maxSpeed              (bits 7 - 14)
+7:  maxSpeed              (bits 15 - 21)
+8:  maxSpeed              (bits 22 - 28)
+9:  maxSpeed              (bits 29 - 32)
+10  csPin                 [optional] (0-127) The chip select pin number
+11: END_SYSEX
+```
+
+#### bitOrder
+```
+LSBFIRST = 0
+MSBFIRST = 1
+```
+
+#### dataMode
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+#### maxSpeed
+
+The maximum speed of communication with the SPI device. For a SPI device rated up to
+5 MHz, use 5000000.
+
+
+### SPI_END_TRANSACTION
+
+Signals the end of a transaction. Must be sent before starting communication
+with another SPI device.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_END_TRANSACTION   (0x02)
+3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### SPI_TRANSFER
```

Comment:
I'm debating whether or not I should break this up into 3 separate commands:

1. SPI_WRITE (write only)
2. SPI_READ (read only, but with option to write 1 or more bytes before reading)
3. SPI_FULL_DUPLEX (for each byte written a byte is read simultaneously)

One thing that would be convenient that the current proposal does not have would be the ability to write a few bytes when doing a read (to accomplish a register read for example). That could look like this if `SPI_TRANSFER` is broken into 3 separate commands as described above:

```
0:  START_SYSEX
1:  SPI_DATA              (0x68)
2:  SPI_READ              (TBD)
3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
4:  pinOptions:           bit 0: CS_DISABLE
                          bit 1: CS_START_ONLY
                          bit 2: CS_END_ONLY
                          bit 3: CS_ACTIVE_EDGE
                          bit 4: CS_TOGGLE
5.  numBytes              (number of bytes to read)
// optional bytes to write before reading
6:  writeData 0           (bits 0-6)
7:  writeData 1           (bit 7) // or inverted depending on bitOrder setting
...
6|N:  END_SYSEX
```